### PR TITLE
node related optimizations

### PR DESCRIPTION
**Pull request by YunHsiao**
Re: cocos-creator/3d-tasks#7487

Changelog:
 * non-recursive invalidate children
 * use Uint32Arrays for hasChangedFlags
* cache pass pool data in JS side

### CODE REVIEW

File: cocos/core/scene-graph/node.ts [hunk 1]
```diff
@@ -533,15 +587,23 @@ export class Node extends BaseNode {
...
+            const hasChangedFlags = cur.hasChangedFlags;
+            if (cur.isValid && (cur._dirtyFlags & hasChangedFlags & dirtyBit) !== dirtyBit) {
+                cur._dirtyFlags |= dirtyBit;
+                cur.hasChangedFlags = hasChangedFlags | dirtyBit;
+                const children = cur._children;
+                const len = children.length;
+                for (let j = 0; j < len; ++j) array_a[++i] = children[j];
+            }
+            dirtyBit = childDirtyBit;
         }
```
Comment (pandamicro) [inline cocos/core/scene-graph/node.ts:607]: better set `array_a.length = 0` here, avoid potential leak

File: cocos/core/scene-graph/node.ts [hunk 2]
```diff
@@ -1111,8 +1173,7 @@ export class Node extends BaseNode {
      * @zh
      * 清除所有节点的脏标记。
      */
-    public static clearBooks () {
-        if (JSB) bookOfChange.forEach((v, k, m) => { if (k.isValid) k.hasChangedFlags = TransformBit.NONE; });
```
Comment (pandamicro) [inline cocos/core/scene-graph/node.ts:1115]: 👍  good job on removing this logic

**Review by pandamicro — CHANGES_REQUESTED** [review]
webgl2-commands is polluted and not possible to review, please explain what have been done in this file

**Review by pandamicro — APPROVED** [review]

**Comment by minggo** [comment]
Any logic to free the pool?

File: cocos/core/renderer/core/pass.ts [hunk 3]
```diff
@@ -693,20 +672,36 @@ export class Pass {
...
     get blocks (): Float32Array[] { return this._blocks; }
+    get rootBufferDirty (): boolean { return this._rootBufferDirty; }
     // states
     get handle (): PassHandle { return this._handle; }
-    get priority (): RenderPriority { return PassPool.get(this._handle, PassView.PRIORITY); }
-    get primitive (): PrimitiveMode { return PassPool.get(this._handle, PassView.PRIMITIVE); }
-    get stage (): RenderPassStage { return PassPool.get(this._handle, PassView.STAGE); }
-    get phase (): number { return PassPool.get(this._handle, PassView.PHASE); }
+
+    set priority (val: RenderPriority) { this._priority = val; PassPool.set(this._handle, PassView.PRIORITY, val); }
```
Comment (minggo) [inline cocos/core/renderer/core/pass.ts:679]: Is it better to record data into PassPool only in JSB?

**Pull request merged**